### PR TITLE
537 add invoices

### LINKED ISSUE #537: Handle Stripe invoice.payment_succeeded webhook event
# Problem

* We should handle the `invoice.payment_succeeded` event
* We store the invoice in a local table
  
```Elixir
field :amount_due
field :application_fee

field :id_from_stripe
field :charge_id_from_stripe
field :customer_id_from_stripe
field :subscription_id_from_stripe

belongs_to :user # can figure it out through customer
belongs_to :subscription

# any other fields that make sense
```

* we then track the event. tracking should contain a user id

### PR DESCRIPTION
**Pull request by begedin**
# What's in this PR?

Work in progress for #537

TODO:

* tests
* implement adapter mapping
* implement figuring out user from customer in adapter
* implement tracking properties
* write do_track method better

## References
Fixes #537

### CODE REVIEW

**Comment by joshsmith** [comment]
This is a little janky. I would prefer better ways of handling all the related resources, but I couldn't think of any. It works, though.

We also should probably write a better set of property tracking for invoices.

File: lib/code_corps/stripe_service/adapters/stripe_invoice.ex [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  alias CodeCorps.{Repo, StripeConnectCustomer, StripeConnectSubscription}
+
+  import CodeCorps.MapUtils, only: [keys_to_string: 1]
+  import CodeCorps.StripeService.Util, only: [transform_map: 2]
+
+  # Mapping of stripe record attributes to locally stored attributes
+  # Format is {:local_key, [:nesting, :of, :stripe, :keys]}
+  @stripe_mapping [
+    {:id_from_stripe, [:id]},
+    {:amount, [:amount]},
```
Comment (begedin) [inline lib/code_corps/stripe_service/adapters/stripe_invoice.ex]: I'm not seeing `:amount` in the API docs? Where do we get it from?

File: lib/code_corps/stripe_service/stripe_invoice_service.ex [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  alias CodeCorps.{Repo, StripeConnectAccount, StripeConnectCustomer, StripeInvoice}
+  alias CodeCorps.StripeService.Adapters.StripeInvoiceAdapter
+
+  @api Application.get_env(:code_corps, :stripe)
+
+  @spec create(binary, binary) :: {:ok, %StripeInvoice{}} | {:error, %Ecto.Changeset{}}
+  def create(invoice_id_from_stripe, customer_id_from_stripe) do
+    with account_id <- get_connect_account(customer_id_from_stripe),
+      {:ok, %Stripe.Invoice{} = invoice} <- @api.Invoice.retrieve(invoice_id_from_stripe, connect_account: account_id),
+         {:ok, params} <- StripeInvoiceAdapter.to_params(invoice)
```
Comment (begedin) [inline lib/code_corps/stripe_service/stripe_invoice_service.ex:12]: Some weird indentation issue here.

**Review by begedin — COMMENTED** [review]
Generally looks good, just confused about the `:amount` field.

Also, I can't approve or request changes, since I'm the pull request author 👍

File: lib/code_corps/stripe_service/adapters/stripe_invoice.ex (hunk 1 above)
Comment (joshsmith) [reply lib/code_corps/stripe_service/adapters/stripe_invoice.ex]: You're right. Idk how I got it.

Comment (joshsmith) [reply lib/code_corps/stripe_service/adapters/stripe_invoice.ex]: Need to change in stripity.

**Review by joshsmith — APPROVED** [review]

**Pull request merged**